### PR TITLE
Expose token-specialized attention compilation and AOT export

### PR DESCRIPTION
**Pull request by TidalPaladin**
## Motivation

Token-specialized attention must avoid TorchInductor's large-batch dynamic-backward scheduler failure while still giving end users control over compilation policy. The existing fixed static workaround prevented users from choosing dynamic compilation, enforcing known static batches, or opting into CUDA GEMM autotuning. The model also returned a custom `ViTFeatures` container that could not be exported and packaged directly without an output wrapper.

The existing direct `vit.attention.attention_token_specialized_qkv_packed` entry point must retain the same automatic large-batch fallback as `SelfAttention`.

## Solution

Expose isolated automatic, dynamic, static, and static-max-autotune attention paths through `ViTConfig`. Validate and propagate exact static batch policies, bypass runtime `torch.compile` dispatch during export, and register `ViTFeatures` as a stable pytree for direct `torch.export` and AOTInductor inference packaging. Route the direct functional API through the same automatic dispatcher so gradient-enabled batches of at least 512 retain the static fallback.

## Changes

- Add the public `TokenSpecializedAttentionCompileMode` type and config fields for compile mode and static batch allowlists.
- Preserve `auto` as the default, including the isolated large-batch training fallback, while keeping forced compiler modes on distinct code objects.
- Preserve the direct token-specialized attention API while routing it through the same automatic compile policy.
- Reject invalid, duplicate, empty, or silently unused compile settings with actionable errors.
- Inline functional token-specialized attention during export and preserve `ViTFeatures` plus token metadata across save/load and AOTInductor packaging.
- Extend `vit-component-benchmark` with production compile modes and a benchmark-only eager diagnostic.
- Record the CUDA performance study and keep compile mode out of case IDs for direct comparisons.
- Document Python/YAML configuration, cache behavior, CUDA autotuning cold starts, export boundaries, and the mode support matrix.
- Attribute the global/visual pathway to Marouani et al.'s [CLS]-patch specialization and identify this repository's register-token and configurable-depth adaptations.

### Configuration example

```python
config = ViTConfig(
    # model and token-specialization fields...
    token_specialized_attention_compile_mode="static",
    token_specialized_attention_static_batch_sizes=(128, 256, 512),
)
```

## Performance

On the retained RTX 3090 BF16 forward/backward workload at `B=512, S=40, D=384`:

| Mode | Mean latency | Relative to `auto` |
| --- | ---: | ---: |
| `auto` | 3.995 ms | baseline |
| `dynamic` | 4.819 ms | 20.6% slower |
| `static` | 4.017 ms | 0.6% slower |
| `static_max_autotune` | 3.859 ms | 3.4% faster |

The public `auto` path is 0.34% slower than the retained 3.981 ms pre-API baseline, within the 5% regression limit. Compilation latency is reported separately; the first max-autotune kernel search took about 24 seconds.

## Test plan

- [x] Focused direct-API regression test, observed failing before the fix and passing after it
- [x] `make quality`
- [x] `make types`
- [x] `make test-ci` (660 passed, 3 skipped; 94% coverage)
- [x] `make test-compile-cpu` (7 passed)
- [x] `make test-compile-cuda` (8 passed, including direct large-batch forward/backward)
- [x] `make check` (961 passed, 5 skipped; 94% coverage)
- [x] Fresh-process public-mode CUDA benchmark trials and dynamic-shape cache sweep

## Test suite changes

- [x] No tests were removed.
- [x] Expanded token-specialization tests cover config validation, YAML normalization, factory propagation, allowlists, stochastic-depth batch semantics, mode dispatch, and export bypass.
- [x] Added a regression proving the direct functional API selects the static fallback for large gradient-enabled batches.
- [x] Added isolated CPU and CUDA compiler checks for every public mode and varied dynamic shapes.
- [x] Strengthened the CUDA regression to compare eager and compiled inference outputs, training outputs, and input/parameter gradients.
- [x] Added direct `torch.export` save/load and AOTInductor package/load parity tests for every mode.
- [x] Extended benchmark parser, artifact, case-ID, and eager diagnostic coverage.

Generated with Codex

### CODE REVIEW

**Comment by TidalPaladin** [comment]
@codex

File: vit/attention.py [hunk 1]
```diff
@@ -371,6 +407,59 @@ def attention_token_specialized_qkv_packed(
     return F.dropout(output, p=dropout, training=training, inplace=True)
 
 
+# Keep ordinary calls dynamic so valid shape changes do not exhaust the full-graph cache.
+attention_token_specialized_qkv_packed = torch.compile(fullgraph=True)(_attention_token_specialized_qkv_packed_impl)
```
[the remark this answers is not in the document]
Comment (TidalPaladin) [reply vit/attention.py]: Fixed in `7b799e9`. The direct functional API is now a compatibility wrapper around the same automatic dispatcher used by `SelfAttention`, so gradient-enabled batches of at least 512 select the isolated static implementation. A focused regression was red before the fix and green afterward, and the CUDA compile suite now exercises this direct large-batch forward/backward path.
